pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -1481,6 +1485,32 @@ func (vs *vSphereVMProvider) restoreVMSpecFromSnapshot(
 	return nil
 }
 
+// unmarshalAndConvertVMFromYAML unmarshals VM YAML from snapshot and converts it
+// to the latest version of VirtualMachine.
+func (vs *vSphereVMProvider) unmarshalAndConvertVMFromYAML(
+	vmCtx pkgctx.VirtualMachineContext,
+	vmYAML string) (*vmopv1.VirtualMachine, error) {
+
+	decUnstructured := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+
+	unstructuredObj := &unstructured.Unstructured{}
+	if _, _, err := decUnstructured.Decode([]byte(vmYAML), nil, unstructuredObj); err != nil {
+		vmCtx.Logger.Error(err, "failed to decode VM YAML in to Unstructured object")
+		return nil, err
+	}
+
+	// Convert it to newest version of VirtualMachine.
+	vm := &vmopv1.VirtualMachine{}
```

Comment:
Would the current VM object be always of the latest version? AFAIU, `vmopv1` will always point to the latest version.

Reply:
Yes, since we have storage: true for the latest version in the CRD, the VM CR got from APIServer will always be at the latest version